pkg/convert/pprof/streaming/parser_streaming_writebatch.go

```diff
@@ -0,0 +1,288 @@
+package streaming
+
+import (
+	"context"
+	"fmt"
+	"github.com/cespare/xxhash/v2"
+	"github.com/pyroscope-io/pyroscope/pkg/stackbuilder"
+	"github.com/pyroscope-io/pyroscope/pkg/util/arenahelper"
+	"reflect"
+	"runtime/debug"
+	"time"
+	"unsafe"
+)
+
+type ParseWriteBatchInput struct {
+	Context            context.Context
+	StartTime, EndTime time.Time
+	Profile, Previous  []byte
+	WriteBatchFactory  stackbuilder.WriteBatchFactory
+}
+
+func (p *VTStreamingParser) ParseWithWriteBatch(input ParseWriteBatchInput) (err error) {
+	defer func() {
+		if recover() != nil {
+			err = fmt.Errorf(fmt.Sprintf("parse panic %s", debug.Stack()))
+		}
+	}()
+	p.startTime = input.StartTime
+	p.endTime = input.EndTime
+	p.ctx = input.Context
+	p.wbf = input.WriteBatchFactory
+	p.cumulative = input.Previous != nil
+	p.cumulativeOnly = input.Previous != nil
+	p.wbCache.cumulative = p.cumulative
+	if input.Previous != nil {
+		err = p.parseWB(input.Previous, true)
+	}
+	if err == nil {
+		p.cumulativeOnly = false
+		err = p.parseWB(input.Profile, false)
+	}
+	p.ctx = nil
+	p.wbf = nil
+	return nil
+}
+
+func (p *VTStreamingParser) parseWB(profile []byte, prev bool) (err error) {
+	return decompress(profile, func(profile []byte) error {
+		p.profile = profile
+		p.prev = prev
+		err := p.parseDecompressedWB()
+		p.profile = nil
+		return err
+	})
+}
+
+func (p *VTStreamingParser) parseDecompressedWB() (err error) {
+	if err = p.countStructs(); err != nil {
+		return err
+	}
+	if err = p.parseFunctionsAndLocations(); err != nil {
+		return err
+	}
+	if !p.haveKnownSampleTypes() {
+		return nil
+	}
+	err = p.UnmarshalVTProfile(p.profile, opFlagParseSamplesWriteBatch)
+	if !p.prev {
+		p.wbCache.flush()
+	}
+	return err
+}
+
+func (p *VTStreamingParser) parseSampleWB(buffer []byte) error {
+	p.tmpSample.reset(p.arena)
+	err := p.tmpSample.UnmarshalSampleVT(buffer, p.arena)
+	if err != nil {
+		return err
+	}
+
+	for i := len(p.tmpSample.tmpStackLoc) - 1; i >= 0; i-- {
+		err = p.addStackLocation(p.tmpSample.tmpStackLoc[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	cont := p.mergeCumulativeWB()
+	if !cont {
+		return nil
+	}
+	p.appendWB()
+	return nil
+}
+
+func (p *VTStreamingParser) appendWB() {
+	for _, vi := range p.indexes {
+		v := uint64(p.tmpSample.tmpValues[vi])
+		if v == 0 {
+			continue
+		}
+		wb := p.wbCache.getWriteBatch(p, vi)
+		if wb == nil {
+			continue
+		}
+
+		sb := wb.wb.StackBuilder()
+		sb.Reset()
+		for _, frame := range p.tmpSample.tmpStack {
+			sb.Push(frame)
+		}
+		stackID := sb.Build()
```

Comment:
Depending on the implementation, `Push` and `Build` may be quite expensive and we may want to minimize the number of calls. I think we can optimise it in feature(!): instead of iterating over frames, we can build an intermediate tree, with a node defined as follows:
```go
type node struct {
  loc    uint32 
  labels []uint64 // each entry points to the labels set (high 32) and offset in the value array (low 32); number of values is known
  values []int64  // groups of values (value idx within a group matches sample type index)
  // ...
}
```

Then, we can traverse the tree and build each frame and stack only once. The reason is that one of our popular "worst" cases assumes lots of unique label sets (e.g exemplars) pointing the same stacks. It _may_ also facilitate diff of cumulative profiles.